FWCore/Services/plugins/StallMonitor.cc

```diff
@@ -0,0 +1,403 @@
+  };
+
+  //===============================================================
+  // Message-assembly utilities
+  template <typename T>
+  std::enable_if_t<std::is_integral<T>::value>
+  concatenate(std::ostream& os, T const t)
+  {
+    os << ' ' << t;
+  }
+
+  template <typename H, typename... T>
+  std::enable_if_t<std::is_integral<H>::value>
+  concatenate(std::ostream& os, H const h, T const... t)
+  {
+    os << ' ' << h;
+    concatenate(os, t...);
+  }
+
+  enum class step : char { preEvent = 'E',
+                           postModuleEventPrefetching = 'p',
+                           preModuleEvent = 'M',
+                           preEventReadFromSource = 'R',
+                           postEventReadFromSource = 'r',
+                           postModuleEvent = 'm' ,
+                           postEvent = 'e'};
+
+  std::ostream& operator<<(std::ostream& os, step const s)
+  {
+    os << static_cast<std::underlying_type_t<step>>(s);
+    return os;
+  }
+
+  template <step S, typename... ARGS>
+  std::string assembleMessage(ARGS const... args)
+  {
+    std::ostringstream oss;
+    oss << S;
+    concatenate(oss, args...);
+    oss << '\n';
+    return oss.str();
+  }
+
+}
+
+namespace edm {
+  namespace service {
+
+    class StallMonitor {
+    public:
+      StallMonitor(ParameterSet const&, ActivityRegistry&);
+      static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+    private:
+
+      void preModuleConstruction(edm::ModuleDescription const&);
+      void postBeginJob();
+      void preEvent(StreamContext const&);
+      void preModuleEvent(StreamContext const&, ModuleCallingContext const&);
+      void postModuleEventPrefetching(StreamContext const&, ModuleCallingContext const&);
+      void preEventReadFromSource(StreamContext const&, ModuleCallingContext const&);
+      void postEventReadFromSource(StreamContext const&, ModuleCallingContext const&);
+      void postModuleEvent(StreamContext const&, ModuleCallingContext const&);
+      void postEvent(StreamContext const&);
+      void postEndJob();
+
+      ThreadSafeOutputFileStream file_;
+      bool validFile_; // Separate data member from file to improve efficiency.
+      std::chrono::milliseconds const stallThreshold_;
+      decltype(now()) beginTime_ {};
+
+      // There can be multiple modules per stream.  Therefore, we need
+      // the combination of StreamID and ModuleID to correctly track
+      // stalling information.  We use tbb::concurrent_unordered_map
+      // for this purpose.
+      using StreamID_value = decltype(std::declval<StreamID>().value());
+      using ModuleID = decltype(std::declval<ModuleDescription>().id());
+      tbb::concurrent_unordered_map<std::pair<StreamID_value,ModuleID>, decltype(beginTime_)> stallStart_ {};
+
+      std::vector<std::string> moduleLabels_ {};
+      std::vector<StallStatistics> moduleStats_ {};
+    };
+
+  }
+
+}
+
+namespace {
+  constexpr char const* filename_default {""};
```

Comment:
An `auto* const` construct only makes sense in a run-time environment.  Since `filename_default` is a compile-time constant (via `constexpr`), the extra `const` is unnecessary, even if permitted by the compiler.